### PR TITLE
[#1838] feat(client): Add Java client support for fileset

### PR DESCRIPTION
**Pull request by coolderli**
### What changes were proposed in this pull request?
Add Java client support for fileset

### Why are the changes needed?

Fix: #1838 

### Does this PR introduce _any_ user-facing change?
- Add Java client support for fileset

### How was this patch tested?

add ut. `com.datastrato.gravitino.client.TestFilesetCatalog`

### CODE REVIEW

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * BaseCatalogDTO is the base class for all the catalog DTOs. It provides the common methods for
+ * managing schemas in a catalog.
+ */
+public class BaseCatalogDTO extends CatalogDTO implements SupportsSchemas {
```
Comment (qqqttt123) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java:39]: We should expose the class `xxxDTO`. The class DTO is only used for transportation between the client and the server.

Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java:39]: How about renaming it to BaseCatalog?

Comment (qqqttt123) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java:39]: I prefer SchemaCatalog for this class. Maybe we can heard from other people's suggestion.

Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: I rename it to `GravitinoCatalog`, it's similar to `GravitinoMetaLake`. What do you think?
```
public class GravitinoMetaLake extends MetalakeDTO implements SupportsCatalogs
```

```
public class GravitinoCatalog extends CatalogDTO implements SupportsSchemas
```

Comment (qqqttt123) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: OK for me. This seems better than BaseCatalog.

Comment (qqqttt123) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: Actually, I have some concern about the name. If we use this class, it means that the other catalogs should also have the concept of schema in the future.

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Fileset catalog is a catalog implementation that supports fileset like metadata operations, for
+ * example, schemas and filesets list, creation, update and deletion. A Fileset catalog is under the
+ * metalake.
+ */
+public class GravitinoFilesetCatalog extends GravitinoCatalog implements FilesetCatalog {
```
Comment (qqqttt123) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:36]: For me, I prefer `FileCatalog`. Because the value of enum is `FILE, RELATIONAL,STREAM`.  We may hear from other people, too. cc @jerryshao @yuqi1129 @mchades

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java (hunk 1 above)
Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: If all catalogs have the concept of schema, what's the disadvantage?

Comment (qqqttt123) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: > If all catalogs have the concept of schema, what's the disadvantage?

I worry whether all catalogs need it. If some catalogs have already mature concepts, we add a concept of schema, it may be weird for the catalog.

Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: > > If all catalogs have the concept of schema, what's the disadvantage?
> 
> I worry whether all catalogs need it. If some catalogs have already mature concepts, we add a concept of schema, it may be weird for the catalog.

It sounds meaningful. I'm not very sure about this. Do we need to have our own standards？ If we have a schema, It's convenient to split the resource. If some resource doesn't need a schema, it can use the same schema by default.

**Comment by coolderli** [comment]
@jerryshao @yuqi1129 @mchades The pipeline had passed. Please take a review when you have time. Thanks.

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java (hunk 2 above)
Comment (yuqi1129) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:36]: I have no preference, both options seem fine to me.

**Comment by YxAc** [comment]
1）Whether all catalogs need schema
- I prefer YES, because schema can make it easy to manage (schema equivalent to database), especially files

2）About Naming
- If all catalogs need schema, the 'SupportsSchemas' can be a base class, doesn't have to be an interface? such as naming 'SchemaCatalog'
- 'GravitinoFilesetCatalog', i prefer naming 'UnstructuredCatalog'

**Comment by qqqttt123** [comment]
> 1）Whether all catalogs need schema
> 
> * I prefer YES, because schema can make it easy to manage (schema equivalent to database), especially files
> 
> 2）About Naming
> 
> * If all catalogs need schema, the 'SupportsSchemas' can be a base class, doesn't have to be an interface? such as naming 'SchemaCatalog'
> * 'GravitinoFilesetCatalog', i prefer naming 'UnstructuredCatalog'
> 
> @jerryshao @coolderli @qqqttt123

Is `UnstructuredCatalog` a good name if we have model catalogs?

**Comment by YxAc** [comment]
> > 1）Whether all catalogs need schema
> > 
> > * I prefer YES, because schema can make it easy to manage (schema equivalent to database), especially files
> > 
> > 2）About Naming
> > 
> > * If all catalogs need schema, the 'SupportsSchemas' can be a base class, doesn't have to be an interface? such as naming 'SchemaCatalog'
> > * 'GravitinoFilesetCatalog', i prefer naming 'UnstructuredCatalog'
> > 
> > @jerryshao @coolderli @qqqttt123
> 
> Is `UnstructuredCatalog` a good name if we have model catalogs?

i think model catalog can extends unstructured catalog, we can discuss this question on DS-Xiaomi regular meeting tomorrow?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java (hunk 2 above)
Comment (mchades) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:36]: How about changing the enum value from `FILE` to `FILE_SET`?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+   * @return The fileset metadata.
+   * @throws NoSuchFilesetException If the fileset does not exist.
+   */
+  @Override
+  public Fileset loadFileset(NameIdentifier ident) throws NoSuchFilesetException {
+    NameIdentifier.checkFileset(ident);
+
+    FilesetResponse resp =
+        restClient.get(
+            formatFilesetRequestPath(ident.namespace()) + "/" + ident.name(),
```
Comment (mchades) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:89]: Do you need to encode the `ident.name()`, in case there are some special character

Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:89]: @mchades Should we validate the special character to avoid using a special character? The `fileset` is managed by gravitino.

Comment (mchades) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:89]: I think it is difficult to enumerate the special characters for validating, using `URLEncoder`  is enough here.

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java (hunk 2 above)
Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:39]: Rename it to FILESET

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java [hunk 4]
```diff
@@ -210,8 +214,23 @@ static String formatFilesetRequestPath(Namespace ns) {
         .toString();
   }
 
+  /** Encode the identifier to be used in the URL. */
+  static String encoder(String ident) {
```
Comment (mchades) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:219]: The method `com.datastrato.gravitino.rest.RESTUtils#encodeString` can be reused. 

BTW, can you also help to replace the method `URLEncoder.encode` in `com.datastrato.gravitino.client.RelationalTable#formatPartitionRequestPath`

Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java]: done

**Comment by coolderli** [comment]
@mchades I found the pipeline is crashed for Javadoc. But I can't figure out the reason. Can you help me about it? 

https://github.com/datastrato/gravitino/actions/runs/7783974587/job/21223570429?pr=2039#step:5:543

**Comment by yuqi1129** [comment]
> @mchades I found the pipeline is crashed for Javadoc. But I can't figure out the reason. Can you help me about it?
> 
> [datastrato/gravitino/actions/runs/7783974587/job/21223570429?pr=2039#step:5:543](https://github.com/datastrato/gravitino/actions/runs/7783974587/job/21223570429?pr=2039#step:5:543)

You need to address all warning informations when building java doc.

**Comment by qqqttt123** [comment]
Error logs in the pipeline
```
/home/runner/work/gravitino/gravitino/clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:223: warning: no @return

  public static Builder builder() {
                        ^
> Task :clients:client-java:javadoc
/home/runner/work/gravitino/gravitino/clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalCatalog.java:263: warning: no @return
  public static Builder builder() {
                        ^
```

**Comment by coolderli** [comment]
@jerryshao @yuqi1129 @qqqttt123 @mchades I have fixed the pipeline. Please review it again when you have time. Thanks.

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java (hunk 1 above)
Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: @jerryshao @qqqttt123 What about renaming it to `SchemaCatalog`?

Comment (qqqttt123) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: > @jerryshao @qqqttt123 What about renaming it to `SchemaCatalog`?

I'm ok.

Comment (jerryshao) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseCatalogDTO.java]: @coolderli sorry for late response. I'm busy on 0.4.0 release. I will give an detailed review today and back to you the  feedbacks.

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoCatalog.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Gravitino Catalog is the base class for all the catalog. It provides the common methods for
+ * managing schemas in a catalog. With {@link GravitinoCatalog}, users can list, create, load, alter
+ * and drop a schema with specified identifier.
+ */
+public class GravitinoCatalog extends CatalogDTO implements SupportsSchemas {
```
Comment (jerryshao) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoCatalog.java:40]: I was thinking of making this as an abstract class with the name "BaseSchemaCatalog"

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Fileset catalog is a catalog implementation that supports fileset like metadata operations, for
+ * example, schemas and filesets list, creation, update and deletion. A Fileset catalog is under the
+ * metalake.
+ */
+public class GravitinoFilesetCatalog extends GravitinoCatalog implements FilesetCatalog {
```
Comment (jerryshao) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:37]: Adding "Gravitino" to the class name seems a bit weird, can we figure out a better name?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+  /** An empty fileset catalog. */
+  public static final GravitinoFilesetCatalog EMPTY = GravitinoFilesetCatalog.builder().build();
```
Comment (jerryshao) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:39]: What is the usage of this static variable?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalTable.java [hunk 8]
```diff
@@ -247,6 +247,6 @@ public SupportsPartitions supportPartitions() throws UnsupportedOperationExcepti
   @VisibleForTesting
   @SneakyThrows // Encode charset is fixed to UTF-8, so this is safe.
   protected static String formatPartitionRequestPath(String prefix, String partitionName) {
-    return prefix + "/" + URLEncoder.encode(partitionName, "UTF-8");
+    return prefix + "/" + RESTUtils.encodeString(partitionName);
```
Comment (jerryshao) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalTable.java:250]: What is the purpose of this change?

File: clients/client-java/src/test/java/com/datastrato/gravitino/client/TestBase.java [hunk 9]
```diff
@@ -77,4 +77,8 @@ protected static <T, R> void buildMockResource(
       throws JsonProcessingException {
     buildMockResource(method, path, Collections.emptyMap(), reqBody, respBody, statusCode);
   }
+
+  protected static String withSlash(String path) {
+    return "/" + path;
+  }
```
Comment (jerryshao) [inline clients/client-java/src/test/java/com/datastrato/gravitino/client/TestBase.java:83]: Seems it is only used in test, why do we move here?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java (hunk 6 above)
Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:37]: @jerryshao The name `FilesetCatalog` is the same with `com.datastrato.gravitino.file.FilesetCatalog`. If we do not rename `FILE` to `FILESET` in the enum, we can call it `FileCatalog`. Do you have any advice?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java (hunk 7 above)
Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/GravitinoFilesetCatalog.java:39]: @jerryshao 
I think it's useless. Just make it the same with https://github.com/datastrato/gravitino/pull/1663. I will remove it.

File: clients/client-java/src/test/java/com/datastrato/gravitino/client/TestBase.java (hunk 9 above)
Comment (coolderli) [reply clients/client-java/src/test/java/com/datastrato/gravitino/client/TestBase.java:83]: I moved it to TestBase so we reuse it on the subclass. We can indeed add one on tests.

**Comment by coolderli** [comment]
@jerryshao @qqqttt123 @mchades @yuqi1129 Could you review it again when you have time? Thanks.

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+    Builder withRestClient(RESTClient restClient) {
+      this.restClient = restClient;
+      return this;
+    }
+
+    @Override
+    public FilesetCatalog build() {
+      Preconditions.checkArgument(restClient != null, "restClient must be set");
+      Preconditions.checkArgument(StringUtils.isNotBlank(name), "name must not be null or empty");
```
Comment (qqqttt123) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java:245]: `must not be blank`

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseSchemaCatalog.java [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * BaseSchemaCatalog is the base abstract class for all the catalog with schema. It provides the
+ * common methods for managing schemas in a catalog. With {@link BaseSchemaCatalog}, users can list,
+ * create, load, alter and drop a schema with specified identifier.
+ */
+public abstract class BaseSchemaCatalog extends CatalogDTO implements SupportsSchemas {
```
Comment (qqqttt123) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseSchemaCatalog.java:38]: Must this class be public? Is it default enough?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java (hunk 10 above)
Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java]: Doen

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseSchemaCatalog.java (hunk 11 above)
Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/BaseSchemaCatalog.java]: Down

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+      return this;
+    }
+
+    @Override
+    public FilesetCatalog build() {
+      Preconditions.checkArgument(restClient != null, "restClient must be set");
+      Preconditions.checkArgument(StringUtils.isNotBlank(name), "name must not be blank");
+      Preconditions.checkArgument(type != null, "type must not be null");
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(provider), "provider must not be null or empty");
```
Comment (qqqttt123) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java:248]: `must not be blank`

Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java]: Done

**Comment by qqqttt123** [comment]
LGTM.

**Comment by qqqttt123** [comment]
@jerryshao @yuqi1129 @mchades Could you take an another look?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+   */
+  @Override
+  public boolean dropFileset(NameIdentifier ident) {
+    NameIdentifier.checkFileset(ident);
+
+    DropResponse resp =
+        restClient.delete(
+            formatFilesetRequestPath(ident.namespace())
+                + "/"
+                + RESTUtils.encodeString(ident.name()),
```
Comment (jerryshao) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java:202]: Why do we encode the `name` for fileset here? I found that we don't do this for metalake, catalog, schema, maybe we should unify them all?

Comment (jerryshao) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java:202]: Also if we encode the string here, do we need to decode it on the server side?

Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java:202]: @jerryshao @mchades Actually, I think we can validate the name of a fileset when creating it. Not like a table, the fileset is managed by Gravitino, so we can make sure that the name complies with regulations. But the table, like SAP, we indeed meet some special names that need to be encoded. 

So I think we don't need to encode the name of fileset. Instead, We can verify the legitimacy of the name during creation. What do you think?


The name of SAP table:
image

Comment (coolderli) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java:202]: > Also if we encode the string here, do we need to decode it on the server side?

As for your question, I think it is not necessary to decode it on the server side. Only if we make sure to decode it on the client, it won't have any impact on users. The only thing I can think of is whether we should decode it before writing it to the audit log.

Comment (jerryshao) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/FilesetCatalog.java:202]: From my point, we don't have to encode the fileset name solely here, we can remove this encoding here, and consider this as a whole solution here https://github.com/datastrato/gravitino/issues/1652

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalTable.java [hunk 14]
```diff
@@ -247,6 +247,6 @@ public SupportsPartitions supportPartitions() throws UnsupportedOperationExcepti
   @VisibleForTesting
   @SneakyThrows // Encode charset is fixed to UTF-8, so this is safe.
   protected static String formatPartitionRequestPath(String prefix, String partitionName) {
-    return prefix + "/" + RESTUtils.encodeString(partitionName);
+    return prefix + "/" + URLEncoder.encode(partitionName, "UTF-8");
```
Comment (jerryshao) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalTable.java:250]: @coolderli I think the change is valid, you don't have to revert it.

**Comment by jerryshao** [comment]
Also can please update the branch to rebase to the latest main.

**Review by jerryshao — APPROVED** [review]

File: clients/client-java/src/test/java/com/datastrato/gravitino/client/TestGravitinoMetalake.java [hunk 15]
```diff
@@ -126,7 +126,7 @@ public void testLoadCatalog() throws JsonProcessingException {
         new CatalogDTO.Builder()
             .withName("mock")
             .withComment("comment")
-            .withType(Catalog.Type.FILE)
+            .withType(Catalog.Type.STREAM)
```
Comment (yuqi1129) [inline clients/client-java/src/test/java/com/datastrato/gravitino/client/TestGravitinoMetalake.java:129]: I suggest you add a new `CatalogDTO` instance instead to change it from `STREAM` to `FILE`.

Comment (coolderli) [reply clients/client-java/src/test/java/com/datastrato/gravitino/client/TestGravitinoMetalake.java:129]: @yuqi1129 Does that mean we have to introduce a new `Catalog.Type` like `UNKNOWN`?

Comment (jerryshao) [reply clients/client-java/src/test/java/com/datastrato/gravitino/client/TestGravitinoMetalake.java:129]: I think it is not necessary to add a "UNKONW" type, the current change is OK for me. We can refactor the UT later on.

Comment (yuqi1129) [reply clients/client-java/src/test/java/com/datastrato/gravitino/client/TestGravitinoMetalake.java:129]: OK

**Review by yuqi1129 — APPROVED** [review]
LGTM
